### PR TITLE
ARROW-11340: [C++] Add vcpkg.json manifest to cpp project root

### PR DESCRIPTION
**Pull request by ianmcook**
@kszucs could you review this please? My main purpose in adding this is to improve the experience for Arrow C++ devs using Windows, but I noticed it also relates to your [TODO in #9096](https://github.com/apache/arrow/pull/9096/files#diff-990134cce6657dbbcf95457cf1a56810a7efa1f6cd58ecc27557c7d6ff45b533R67-R68). vcpkg does not have any `requirements.txt`-style package enumeration mechanism, but it supports this JSON manifest as a mechanism of defining dependencies.

In the `vcpkg install` command, you can specify the path to the directory containing this manifest file with `--x-manifest-root` which later will change to `--manifest-root`. See details at https://vcpkg.readthedocs.io/en/stable/specifications/manifests/. 

There are some differences between the packages listed in this manifest versus the packages you listed in the `vcpkg install` commands in #9096 
- This installs `gtest` and `benchmark`
- This installs `boost` instead of separate `boost-filesystem`, `boost-regex`, etc.
- This does not explicitly include the `core` feature of `aws-sdk-cpp` because explicitly including it causes an error, and it gets installed anyway

### CODE REVIEW

**Comment by kszucs** [comment]
Thanks Ian for working on this! 

I have a couple of questions:
- could we define multiple manifest files for different use cases (e.g. for building wheels or having a minimal build)?
- how can one utilize this manifest file (I assume we need to document this in the developer guide)?
- why do we need to add a version string to the manifest (if it is mandatory than we need to update the version bumping scripts used during the release procedure)? 

Could you setup a github actions build to test it?

**Comment by ianmcook** [comment]
Replies inline
> Thanks Ian for working on this!
> 
> I have a couple of questions:
> 
> * could we define multiple manifest files for different use cases (e.g. for building wheels or having a minimal build)?

Yes, definitely. Or would it be better to use this one as the master copy and apply patches to it to create the copies for other use cases?

For safety, we should never name the copies `vcpkg.json` because Visual Studio automatically detects manifests with that file name.

> * how can one utilize this manifest file (I assume we need to document this in the developer guide)? 

I'm working on that in [ARROW-11336](https://issues.apache.org/jira/browse/ARROW-11336)


> * why do we need to add a version string to the manifest (if it is mandatory than we need to update the version bumping scripts used during the release procedure)?

It's mandatory. (The vcpkg docs say so and I confirmed by testing.)

> 
> Could you setup a github actions build to test it?

Sure, I'll work on that

**Comment by ianmcook** [comment]
FYI, the version of vcpkg that is currently preinstalled on the Github Actions Windows images is 2020.11.12 (as noted [here](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md)). This version has a bug (https://github.com/awslabs/aws-c-common/issues/734) that causes the installation of `aws-cpp-sdk` to fail. When running vcpkg in Github Actions on Windows, remove the preinstalled vcpkg and install the newest version from source.

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
@kszucs I added a Crossbow test. It runs a GHA job on a Windows instance. The job installs the current version of vcpkg, installs all the dependencies listed in `vcpkg.json`, runs CMake with many Arrow features enabled, and builds Arrow.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (pitrou) [inline dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:37]: Ninja generally builds much faster than msbuild, unless the latter improved recently.

File: dev/tasks/vcpkg-tests/test-vcpkg-win.yml
Comment (pitrou) [inline dev/tasks/vcpkg-tests/test-vcpkg-win.yml:38]: Would you like to add a comment explaining why?

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (pitrou) [inline dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:20]: I'm curious whether this is necessary. Isn't CMake able to detect the VS install? Or is this for vcpkg?

Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:20]: The CMake that's being used here is the one distributed with Visual Studio. This command adds that CMake to the path and sets some environment variables to make it work correctly. If I omit this, then the `cmake` command below gives `'cmake' is not recognized as an internal or external command,
operable program or batch file.`

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:37]: My intention here was to keep the build as vanilla as possible. The amount time it takes to generate and build is very small compared to the time vcpkg takes to build all the packages, so it won't make much of a proportional difference either way.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (pitrou) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:20]: Ow, I forgot that the vcpkg model is to build everything from source. So this is gonna rebuilt OpenSSL, Boost, gRPC... everytime.

I understand why you didn't enable Gandiva!

File: dev/tasks/vcpkg-tests/test-vcpkg-win.yml
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/test-vcpkg-win.yml:38]: Sure; it's explained in [this comment](https://github.com/apache/arrow/pull/9287#issuecomment-765740804) but I'll add that in a comment here in the code also

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (pitrou) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:37]: I see.

File: dev/tasks/vcpkg-tests/test-vcpkg-win.yml
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/test-vcpkg-win.yml:38]: Added in 5408df1

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (kou) [inline dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:41]: How about using `--parallel %NUMBER_OF_PROCESSOR%` with `cmake --build` instead of specifying `/MP` here?

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (kou) [inline dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:55]: Why should we specify empty prefix?

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (kou) [inline dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:64]: Could you sort this list in alphabetical order?

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (kou) [inline dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:40]: If we enable this, how about running tests not just building?

File: cpp/vcpkg.json
Comment (kou) [inline cpp/vcpkg.json:3]: I'll push a commit to update this automatically in release process.

File: dev/tasks/tasks.yml
Comment (kou) [inline dev/tasks/tasks.yml:1796]: Should we run this job nightly?
We can run this job nightly by adding this to `nightly:` group.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:55]: Thanks for catching that; I'll remove it

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:41]: Sure, I'll change it to that. FYI this `cmake` command was mostly copied from [ci/appveyor-cpp-build.bat](https://github.com/apache/arrow/blob/master/ci/appveyor-cpp-build.bat#L91-L121) with changes only as needed.

File: cpp/vcpkg.json
Comment (kou) [inline cpp/vcpkg.json:17]: Can we remove this?
It seems that we don't run benchmark with the new job in this pull request.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:40]: Yes. Could you point me to an example showing the tests that would be most appropriate here? Thank you

Comment (kou) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:40]: `ctest --output-on-failure --parallel %NUMBER_OF_PROCESSORS --timeout 300` will work.
https://github.com/apache/arrow/blob/master/ci/appveyor-cpp-build.bat#L127 also uses similar command line.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:41]: Upon further investigation, a few sources point to this `/MP` method having better compatibility with Visual Studio than `--parallel`, so I think I will keep it this way

Comment (kou) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:41]: Could you share the sources?

File: cpp/vcpkg.json
Comment (ianmcook) [reply cpp/vcpkg.json:17]: The larger purpose of adding this `vcpkg.json` manifest file is to improve the experience for Windows-based Arrow C++ developers who prefer vcpkg over conda ([ARROW-11336](https://issues.apache.org/jira/browse/ARROW-11336)). We want to give Windows devs two alternative ways to get the same build dependencies: vcpkg or conda. Therefore I want to keep what's here as similar as possible to what's in [ci/conda_env_cpp.yml](https://github.com/apache/arrow/blob/master/ci/conda_env_cpp.yml), which is what we [tell Windows devs to use with conda](https://arrow.apache.org/docs/developers/cpp/windows.html#using-conda-forge-for-build-dependencies). That contains benchmark and gtest, so I think this should too. That way all the latter instructions can be consistent regardless of whether vcpkg or conda was used.

P.S. I can provide the mapping of each conda-forge package to the analogous vcpkg package if you'd like to see that.

Comment (kou) [reply cpp/vcpkg.json:17]: OK. Then how about adding `-DARROW_BUILD_BENCHMARKS=ON`?

File: dev/tasks/tasks.yml
Comment (ianmcook) [reply dev/tasks/tasks.yml:1796]: For now it might be better to run it only when we do release candidate verification.

A next step will be to get an Arrow 3.0.0 "port" (that's the term they use) added to vcpkg, so users can simply `vcpkg install arrow` in their projects that use the Arrow C++ library as a dependency. Currently there is an [old port of Arrow on vcpkg](https://github.com/microsoft/vcpkg/tree/master/ports/arrow) and to date we have not treated vcpkg like a formal release channel. Once that is done, we will want to run nightly tests that install Arrow with vcpkg onto one or more supported platforms.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:41]: - https://gitlab.kitware.com/cmake/cmake/-/issues/20564
- https://blog.kitware.com/cmake-building-with-all-your-cores/ (recommends `/MP`, but it's from 2013)
- https://gitlab.kitware.com/cmake/cmake/-/merge_requests/1962 (seems to show `--parallel` support is somewhat new, and many Windows users run Visual Studio 2017 or newer which I'm not sure supports it)

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

File: dev/tasks/tasks.yml
Comment (kou) [reply dev/tasks/tasks.yml:1796]: OK.
But I don't want to increase tasks that are ran only when release candidate verification. Normally, these tasks are outdated and they increase release cost. I hope that we can go to the next step by the next release.

File: dev/tasks/tasks.yml
Comment (kou) [inline dev/tasks/tasks.yml:1798]: We use `${CI_PLATFORM}.${PLATFORM}.yml` format for template file name:

```suggestion
    template: vcpkg-tests/github.windows.yml
```

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (kou) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:43]: Thanks. I see.
Could you note https://gitlab.kitware.com/cmake/cmake/-/issues/20564 as a comment? We will reconsider this when CMake improves about this.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (kou) [inline dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:67]: Do we need explicit `1` for `exit /B`?
Other `exit /B`s don't specify `1`.

**Review by kou — APPROVED** [review]
+1

We can merge this with some minor improvements and green CI result.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:67]: I think we do (per https://docs.github.com/en/actions/creating-actions/setting-exit-codes-for-actions). Thanks for catching that. I'll add the `1` everywhere.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat]: I suspect that removing this caused the failure on Crossbow: https://github.com/ursacomputing/crossbow/runs/1781750297?check_suite_focus=true#step:4:3807
I'll restore it and see if that fixes the failure

File: dev/tasks/tasks.yml
Comment (ianmcook) [reply dev/tasks/tasks.yml]: Renamed in 4c4247

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
The Crossbow build is failing because of https://github.com/google/benchmark/issues/1046. I will remove Benchmark from the build for now.

**Comment by ianmcook** [comment]
I rebased to resolve a seemingly unrelated test failure

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by kou** [comment]
> Some builds are failing apparently because of [google/benchmark#1046](https://github.com/google/benchmark/issues/1046). I will remove Benchmark from the build for now.

Could you note this as a comment?

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
I believe the **C++ / AMD64 Windows MinGW 64 C++** test failure is random and unrelated to the changes in this PR. It passed before I pushed the last two commits, and those commits only added comments and whitespace.

**Comment by kou** [comment]
You can ignore the failure in this pull request.

File: dev/tasks/vcpkg-tests/github.windows.yml
Comment (kou) [inline dev/tasks/vcpkg-tests/github.windows.yml:63]: We need to set `ARROW_TEST_DATA` and `PARQUET_TEST_DATA` environment variables for test. Sorry for not mentioning them. The following will work:

```yaml
env:
  ARROW_TEST_DATA: ${{ env.GITHUB_WORKSPACE }}\arrow\testing\data
  PARQUET_TEST_DATA: ${{ env.GITHUB_WORKSPACE }}\arrow\cpp\submodules\parquet-testing\data
```

Comment (ianmcook) [reply dev/tasks/vcpkg-tests/github.windows.yml:63]: Thanks. Added in the `.bat` file in 3c3d139

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
@kszucs we're getting two failures in `ctest` when the dependencies are installed with vcpkg:
```
[ RUN      ] TestStatisticsSortOrder/0.MinMax
D:/a/crossbow/crossbow/arrow/cpp/src/parquet/statistics_test.cc(667): error: Expected equality of these values:
  stats_[i].min()
    Which is: "\0\0\0\0"
  cc_metadata->statistics()->EncodeMin()
    Which is: "\x3\0\0\0"
D:/a/crossbow/crossbow/arrow/cpp/src/parquet/statistics_test.cc(835): error: Expected: this->VerifyParquetStats() doesn't generate new fatal failures in the current thread.
  Actual: it does.
[  FAILED  ] TestStatisticsSortOrder/0.MinMax, where TypeParam = struct parquet::PhysicalType<1> (0 ms)
```

```
[ RUN      ] TestStatistic.Int32Extremums
D:/a/crossbow/crossbow/arrow/cpp/src/parquet/statistics_test.cc(897): error: Expected equality of these values:
  stats->min()
    Which is: -2147483648
  expected_min
    Which is: 0
[  FAILED  ] TestStatistic.Int32Extremums (0 ms)
```

Do we need to apply a patch to one or more vcpkg ports to fix this? Is this why you apply [ci/vcpkg/ports.patch](https://github.com/apache/arrow/blob/master/ci/vcpkg/ports.patch) to the vcpkg sources when building wheels?

**Comment by kszucs** [comment]
The issue may lay in the implementation, so I suggest to create follow-up issues in order to investigate it.

**Comment by ianmcook** [comment]
> The issue may lay in the implementation, so I suggest to create follow-up issues in order to investigate it.

Ok, thank you. I added a TODO comment just above the `ctest` call, and I'll open a separate issue to investigate this. Can we merge this PR despite this Crossbow job failing? That way I can use the same Crossbow job to troubleshoot the issue in a separate PR.

**Comment by ianmcook** [comment]
@kszucs After this is merged, there is a slight possibility that we could see unexpected behavior in `ci/docker/python-wheel-windows-vs2017.dockerfile` and `ci/docker/python-wheel-manylinux-201x.dockerfile`. I do not think this will happen, but there's a chance that Visual Studio's CMake will automatically identify the `vcpkg.json` manifest file and attempt to install all the packages according to that manifest instead of using the ones installed by the prior `vcpkg` command. This should not happen because `vcpkg integrate install` has not been run, but in the event that it does happen, the solution is to run `vcpkg integrate remove` and/or to delete `vcpkg.json` before running `cmake`.

**Comment by kszucs** [comment]
@github-actions crossbow submit wheel-manylinux2010-cp36m wheel-windows-cp36m

**Comment by ianmcook** [comment]
@github-actions crossbow submit wheel-manylinux2010-cp36m wheel-windows-cp36m

File: cpp/CMakeSettings.json
Comment (ianmcook) [inline cpp/CMakeSettings.json:21]: For developers who are using Visual Studio 2019 and have vcpkg installed, when the `arrow/cpp` directory is opened in the Visual Studio IDE, it automatically identifies the `vcpkg.json` manifest and begins installing the dependencies listed in it into a subdirectory within the project. This is an undesirable behavior because the developer might want to use conda to install the dependencies or they might have already used vcpkg to install the dependencies outside the project. This `CMakeSettings.json` file sets the CMake variable `VCPKG_MANIFEST_INSTALL` to `OFF`, preventing Visual Studio from performing this automatic action. In [ARROW-11336](https://issues.apache.org/jira/browse/ARROW-11336), I will document this and instruct users to change the value to `ON` if they do wish for vcpkg to automatically install dependencies.

**Comment by ianmcook** [comment]
@github-actions crossbow submit wheel-manylinux2010-cp36m wheel-windows-cp36m

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
I expect the *test-build-vcpkg-win* test to show a failure, for reasons described above

**Comment by ianmcook** [comment]
The **test-build-vcpkg-win** failure is for the same reasons as noted above in https://github.com/apache/arrow/pull/9287#issuecomment-769843489 and should not block this from being merged.

**Comment by ianmcook** [comment]
@github-actions crossbow submit wheel-manylinux2010-cp36m wheel-windows-cp36m

**Comment by ianmcook** [comment]
@github-actions crossbow submit wheel-manylinux2010-cp36m wheel-windows-cp36m

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by ianmcook** [comment]
@github-actions crossbow submit wheel-manylinux2010-cp36m wheel-windows-cp36m

**Comment by ianmcook** [comment]
@kszucs My latest commits in this PR fix the unintended side effects that were occurring because of the addition of the manifest file `cpp/vcpkg.json`:
- The Python wheel builds (which use vcpkg to install dependencies but are _not_ supposed to use the manifest) were failing, because the existence of the manifest was causing the Visual Studio CMake toolchain to behave in some unexpected ways. I have set several variables and options that resolve this.
- When a developer opened the `cpp` directory in the Visual Studio 2019 IDE, it would automatically recognize the manifest and initiate a `vcpkg install` regardless of whether the developer intended to use vcpkg to install dependencies. I added a `CMakeSettings.json` file to disable this behavior.

The **wheel-manylinux2010-cp36m** and **wheel-windows-cp36m** Crossbow builds are now passing

The **test-build-vcpkg-win** failure is for the same reasons as noted above in https://github.com/apache/arrow/pull/9287#issuecomment-769843489 and should not block this from being merged.

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (kszucs) [inline dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:90]: @ianmcook please create a follow-up jira for this

**Review by kszucs — APPROVED** [review]
LGTM, thanks Ian!

File: dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat
Comment (ianmcook) [reply dev/tasks/vcpkg-tests/cpp-build-vcpkg.bat:90]: [ARROW-11675](https://issues.apache.org/jira/browse/ARROW-11675)

**Pull request closed without merging**
